# What Attributes and Genres Appeal to the Most users in the Free to use app market


## This project is going to be focussing on the mobile phone app market specifically that of apps downloaded on the Apple Store and the Google Play Store. 

The project that i am about to complete is going to be looking at data that we have been able to get from both the IOS and Google Play store platforms. We are going to clean the data so that any missing data is erased, duplicate data is retrived and removed and data that doesnt conform with what we need is also not used so that the analysis that i complete is as fair and rational as possible. 

The data that i have used for this project has been sourced from:
- Google play store (https://www.kaggle.com/lava18/google-play-store-apps)
- Apple Store IOS (https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps)

In [3]:
opened_file = open('AppleStore.csv', encoding="utf8")
opened_file2 = open('googleplaystore.csv', encoding="utf8")
from csv import reader
read_file = reader(opened_file)
read_file2 = reader(opened_file2)
a = list(read_file)
g = list(read_file2)
apple_data = a[1:]
google_data=g[1:]

apple_header = a[0]
google_header = g[0]
print(len(google_header))
print(len(apple_header))

13
16


# Opening and Exploring the Data

__Google Play Store Columns__  
App, Category, Rating, Reviews, Size, Installs, Type, Price, Content Rating, Genres, Last Updated, Current Ver Android Ver

__Apple Store IOS Columns__
id, track_name, size_bytes, currency, price, rating_count_tot, rating_count_ver, user_rating, user_rating_ver, ver cont_rating
prime_genre, sup_devices.num, ipadSc_urls.num, lang.num, vpp_lic

As we can see with the above code we have been apple to sperate the column names for each of the data sets so that we can make sure that we are aware of the meaning of all of the columns. This further helps us clean the data as it has now become apparent that some of the columns are not sueful for the task that we have at hand. 

I have identified the columns applicable for the Google Play store as being: 
App, Catergory, Rating, Reviews, Installs, Content Rating. Genre As the columns in the Google Play Store data that most corralate to what we are looking to acheive. This takes us from 13 different columns of data down to 7

I have identified the columns applicable for the ApplePlay store as being: 
track_name, rating_count_tot, user_rating, prime_genre, lang.num

In [4]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))
        
explore_data(apple_data,0,3,True)
explore_data(google_data,0,3,True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7197
Number of columns: 16
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2

# Deleting Wrong Data

Now that this peice of data has been deleted a new data point will occupy this space so to run it twice will ruin the data set that we have ammassed. 

While looking at the discussion for this data it is said that one of the data points is missing some data, due to the fact we want to only use data that is 100% complete i looked into this and as you can see the data point is only showing 12 different columns attached to it meaning that we are missing one peice of data. Due to this i have deletd this data point.

In [5]:
print(google_data[10472])
print(len(google_data[10472]))

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']
12


In [6]:
del google_data[10472]

# Dealing with Duplicate Data

We are now going to look at both of the stores data sets to determine if there are any data points that have been duplicated. To this we are going to loop through the name column for both of the data sets, which is 'track_name' for the IOS and 'App' for the google play store

In [7]:
apple_single = []
apple_dup = []
for row in apple_data:
    name = row[1]
    if name in apple_single:
        apple_dup.append(name)
    else:
        apple_single.append(name)        
print(f'Number of Duplicates in the IOS store is {len(apple_dup)}')


google_single = []
google_dup = []
for row in google_data:
    name = row[0]
    if name in google_single:
        google_dup.append(name)
    else:
        google_single.append(name)        
print(f'Number of Duplicates in the Google Play Store store is {len(google_dup)}')


Number of Duplicates in the IOS store is 2
Number of Duplicates in the Google Play Store store is 1181


As you can see from the above, amount of duplicates that are in the dataset in the google play store far surpass that of the IOS. 

In [8]:
for app in google_data:
    name = app[0]
    if name == 'Facebook':
        print(app)

['Facebook', 'SOCIAL', '4.1', '78158306', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'August 3, 2018', 'Varies with device', 'Varies with device']
['Facebook', 'SOCIAL', '4.1', '78128208', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'August 3, 2018', 'Varies with device', 'Varies with device']


Here is a random example of a duplicate app that we have found in our data set. As we do not want to use the app more then once it is prudent that we remove the duplicate data. One of the columns has a different value to the duplicate data whihc is the amount of user reviews. This leads me to belive that the data point wiht a higher number is the more recent data point and thus is the data poiint that we should be keeping for our analysis. 

In [9]:
reviews_max = {}

for row in google_data:
    name = row[0]
    n_reviews = float(row[3])
    if name in reviews_max and reviews_max[name]<n_reviews:
        reviews_max[name]=n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews
print(len(reviews_max))

print(f'Number of data points: {len(google_data)}')
print(f'Number of duplicate data points: {len(google_dup)}')


9659
Number of data points: 10840
Number of duplicate data points: 1181


In the above peice of code we have take the data from the google play store and made a new dictionary out of the data using the apps name as the Key and the number of reviews as the value. As we can also show the remaining duplicate entires have been removed from the data set with 9659 data points remaining in our dataset.

- We start the process of removing the duplicates by initilizing two empty lists
- We then loop through each interation of the data and isolate the name and reviews 
- If the number of reviews is the same as the value of the dictionary then the entire row for that app is added to the clean data
- To make sure apps with the same amount of reviews dont get through we make sure to add the name o feach app to a second list and requre that the name not already be in our clean data

In [10]:
android_clean = []
already_added = []

for row in google_data:
    name = row[0]
    n_reviews = float(row[3])
    if (n_reviews ==reviews_max[name]) and (name not in already_added):
        android_clean.append(row)    
        already_added.append(name)
len(android_clean) 


9659

# Removing non-English Apps

We want the app data that we analyise to be of apps that are english language only so it is important that we remove apps from the analysis that are from different countries. We can do this by removing any apps that have letters no represented on a english keyboard. We can use the ord function as we know that the english keyboard only has a representation upto ord 127 so anything after this must mean it is not an english app

Due to the fact that emojis are now a very prevalanet addition into the english language it has been decided that because they are outside the range that we are using we will require an app to have more then 3 over 127 valued chars so to include as much relevant data, this is not perfect but it is better the before. 

In [11]:
def english_only(app_name):
    letter = 0
    for a in app_name:
        if ord(a)>127:
            letter+=1
    if letter > 3:
        return False
    else: 
        return True

In [12]:
explore_data(android_clean, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns: 13


Now that we have the function that allows us to check if 3 or more letters outside of the english keyoboard, the data of both app stores has been put into the function are two new lists have been created that only hold the clean data of english language apps. 

In [13]:
andorid_clean_english = []
ios_clean_english = []

for row in android_clean:
    name = row[0]
    
    english = english_only(name)
    if english == True:
        andorid_clean_english.append(row) 
print(len(andorid_clean_english))        
    
for row in apple_data:
    name = row[1]
    
    english = english_only(name)
    if english == True:
        ios_clean_english.append(row) 
print(len(ios_clean_english))         
    

9614
6183


# Isolating the Free Apps 

In this step of the data cleaning process we are going to loop through the apps that we have left and then remove any that are not free of charge. To do this i am going to have to isolate the price and use loops and conditionals to filter out the data that we need.

In [14]:
free_clean_android = []
free_clean_ios = []


for row in andorid_clean_english:
    price = row[7]   
    if price =='0':
        free_clean_android.append(row)
print(len(free_clean_android))

for row in ios_clean_english:
    price = row[4]    
    if price =='0.0':
        free_clean_ios.append(row)
print(len(free_clean_ios))


8864
3222


After the non-free apps were removed from the data we were left with 8864 android apps and 3222 IOS apps

In [18]:
print(google_header, '\n')
print(google_data[65], '\n')
print(apple_data[2])
print(apple_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 

['Check Vehicle Tax', 'AUTO_AND_VEHICLES', '3.9', '295', '2.7M', '10,000+', 'Free', '0', 'Everyone', 'Auto & Vehicles', 'July 30, 2018', '0.2.1', '4.4 and up'] 

['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']
['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


Now that the data has been cleaned it is finally time that we look at the final cleaned data points and find ways in which to analyise it for the purpose that we originally set out for. 
From the columns that i have just pulled up it seems to me that the best data points to use for frequency tables would be 

Google - Category, installs
IOS - prime_genre

In [42]:
def freq_table(dataset,index):
    freq_dict = {}
    count = len(dataset)
    for row in dataset:
        if row[index] in freq_dict:
            freq_dict[row[index]]+=1
        else:
            freq_dict[row[index]]=1
    table_percentage = {}        
    for genre in freq_dict:
        percentage = (freq_dict[genre] /count)*100
        table_percentage[genre] = percentage
    return table_percentage    

In [43]:
android_genre_per= freq_table(free_clean_android,1)
ios_genre_per = freq_table(free_clean_ios,11)

In [46]:
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [47]:
display_table(free_clean_android,1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

In [48]:
display_table(free_clean_ios,11)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665
